src/ast/mod.rs

```diff
@@ -335,7 +335,29 @@ pub enum Command {
         sort: Symbol,
     },
     Sort(Symbol, Option<(Symbol, Vec<Expr>)>),
+    /// Declare an egglog function.
+    /// The function is a datatype when:
+    /// - The output is not a primitive
+    /// - No merge function is provided
+    /// - No default is provided
     Function(FunctionDecl),
```

Comment:
Oh I wasn't aware you could have functions with non primitive return values and custom merge functions...

Reply:
It's currently a very under-tested feature! I plan to use it more in the future though, it's essential for encoding proofs.

Reply:
It can be used to simulate the [`choice`](https://souffle-lang.github.io/choice) construct in Datalog literature by setting `:merge old` (as is used in proofs)